Iteration 5

mikesvis opened the pull request:
Env config + making config simpler

mikesvis committed:
Iteration 1

mikesvis committed:
Iteration 1. Server

mikesvis committed:
Iteration 1. Linter issues

mikesvis committed:
Iteration 1. Change project structure

mikesvis committed:
Iteration 2: Tests + code tweaks (uuid, decoupled error response, etc)

mikesvis committed:
Iteration 3. Router

mikesvis committed:
Iteration 4. Flags + config

mikesvis committed:
Iteration 5. Env config + making config simpler

capsulated commented on internal/app/config/config.go:
```diff
@@ -0,0 +1,71 @@
+	return nil
+}
+
+func (a Address) String() string {
```
А тут специально без указателя ресивер? Если да, то почему?

mikesvis replied:
Видимо не совсем хорошо разобрался в теме с указателями и мной двигала идея защиты от изменения, которая тут непричем. Исправил.

capsulated replied:
Это надо обязательно разобраться) Поэтому вопрос такой оставил. Много уже и видео и статей по этой теме: "передача параметров по указателю или по значению".

capsulated commented on internal/app/config/config.go:
```diff
@@ -0,0 +1,71 @@
+package config
```
`app` пакет, на мой взгляд, лишний. При создании директории мы автоматически создаём пакет - это одно из существенных отличий go от других ЯП. Директория, это что-то типа namespace из других языков

Лучше создать пакеты `config`, `server`, `storage`, а `helpers` разбть бы на два независимых пакета: `urlformat`, `keygen`, так как utils, helpers и т.п -. антипаттерн, он "распухнет" с развитием проекта

mikesvis replied:
Реорганизовал структуру. До последнего момента хотел сделать разные пакеты для `urlformat` и `keygen`, но в виде подпапок в папке `helpers`. Не понял что я сам от этого выиграю и сделал без папки `helpers`.

capsulated commented on internal/app/server/handler.go:
```diff
@@ -0,0 +1,93 @@
+// Обработка Get
+// Получение короткого URL из запроса
+// Поиск в условной "базе" полного URL по сокращенному
+func ServeGet(s storage.StorageURL) http.HandlerFunc {
```
Удобнее описать структурный тип с методами хендлеров. В таком формате удобнее добавлять зависимости (в нашем случае во всех хендлерах параметр хранилки), скажем если добавиться логгер, кэш, трассировка или подобное, придётся добавлять к каждому хендлеру

mikesvis replied:
Сделал насколько хватило ума-разума. По хорошему бы конечно повторить структуру `service` `storage` `api` `model` как показали на вебинаре дабы покрыть интерфейсами где и как-надо, но пока решил оставить близко к своей структуре. Учту на будущее.

capsulated commented on internal/app/server/server.go:
```diff
@@ -0,0 +1,35 @@
+	s = storage.NewStorageURL(make(map[domain.ID]domain.URL))
+}
+
+func ShortRouter() chi.Router {
```
Это фактически конструктор, лучше именовать его `NewRouter() *chi.Mux { ... }` и возвращать указатель на экземпляр, а не интерфейс. Про нейминг интерфейсов [тут](https://go.dev/doc/effective_go#composite_literals)

mikesvis replied:
Про конструкторы понял, но так же элегантно как в примере (в одну строчку) с описанием роутинга не смог придуать как реализовать. Поменял возвращаемый тип.

capsulated replied:
Это вообще не критично, чуть чуть проще будет читать в будущем. Будет кейс, где возвращается интерфейс (абстрактная фабрика). А конструктор в `go` - это просто файбрика: на вход параметры, на выходе - указатель на созданный экземпляр

capsulated commented on internal/app/storage/storage.go:
```diff
@@ -0,0 +1,48 @@
+	"github.com/mikesvis/short/internal/domain"
+)
+
+type StorageURL interface {
```
Интерфейс принято описывать в местах использования, а не реализации

mikesvis replied:
Как говорил выше по хорошему бы структуру видоизменить. Возможно столкнусь в дальнейшем и переделаю. Пока что вынес интерфейс отдельно от реализации storage.

capsulated reviewed (changes requested):
В общем код организован хорошо, но есть некоторые замечания по структуре, см в коментах

mikesvis committed:
Iteration 5. Corrections

1. Address String() by pointer
2. Packages reorganization
3. NewRouter {} *chi.Mux

mikesvis committed:
Iteration 5. Correction pt2. Handler DI & methods

mikesvis committed:
Iteration 5. Correction pt3. Storage interface out from storage realization

capsulated commented on internal/server/handler.go:
```diff
@@ -0,0 +1,93 @@
+// Проверка на валидность URL
+// Запись сокращенного Url в условную "базу" если нет такого ключа
+func (h *handler) ServePost() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
```
Тут уже можно убрать возвращаемую функцию на обычную сигнатуру: 
```go
func (h *handler) ServePost() (w http.ResponseWriter, r *http.Request) { ... }
```

Но в таком случае, "гибкость" кода становиться меньше.
Так что, смотри сам, как лучше

capsulated reviewed (approved):
Да, всё круто!
